include/pybind11/cast.h

```diff
@@ -1637,10 +1637,15 @@ str_attr_accessor object_api<D>::attr_with_type_hint(const char *key) const {
         throw std::runtime_error("__annotations__[\"" + std::string(key) + "\"] was set already.");
     }
 
-    const char *text = make_caster<T>::name.text;
+    static constexpr auto type_name = make_caster<T>::name;
+    PYBIND11_DESCR_CONSTEXPR auto types = decltype(type_name)::types();
```

Comment:
These are very generic names. It'd be great to make them more descriptive. How about
```
static constexpr auto caster_name_field = make_caster<T>::name;
PYBIND11_DESCR_CONSTEXPR auto descr_types = decltype(caster_name_field)::types();
```